src/server/tablet_snap.rs

```diff
@@ -0,0 +1,526 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{
+    convert::TryFrom,
+    fs::{self, File},
+    io::{Read, Write},
+    marker::PhantomData,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+    time::Duration,
+};
+
+use engine_traits::KvEngine;
+use file_system::{IoType, WithIoType};
+use futures::{
+    future::{Future, TryFutureExt},
+    sink::{Sink, SinkExt},
+    stream::{Stream, StreamExt, TryStreamExt},
+};
+use grpcio::{
+    self, ChannelBuilder, ClientStreamingSink, Environment, RequestStream, RpcStatus,
+    RpcStatusCode, WriteFlags,
+};
+use kvproto::{
+    raft_serverpb::{Done, RaftMessage, RaftSnapshotData, SnapshotChunk},
+    tikvpb::TikvClient,
+};
+use protobuf::Message;
+use raftstore::{
+    router::RaftStoreRouter,
+    store::snap::{TabletSnapKey, TabletSnapManager},
+};
+use security::SecurityManager;
+use tikv_util::{
+    config::{Tracker, VersionTrack},
+    time::Instant,
+    worker::Runnable,
+};
+use tokio::runtime::{Builder as RuntimeBuilder, Runtime};
+
+use super::{metrics::*, snap::Task, Config, Error, Result};
+use crate::tikv_util::{sys::thread::ThreadBuildWrapper, time::Limiter};
+
+pub type Callback = Box<dyn FnOnce(Result<()>) + Send>;
+
+const DEFAULT_POOL_SIZE: usize = 4;
+
+const SNAP_CHUNK_LEN: usize = 1024 * 1024;
+
+struct RecvTabletSnapContext {
+    key: TabletSnapKey,
+    raft_msg: RaftMessage,
+    io_type: IoType,
+    start: Instant,
+}
+
+impl RecvTabletSnapContext {
+    fn new(head_chunk: Option<SnapshotChunk>) -> Result<Self> {
+        // head_chunk is None means the stream is empty.
+        let mut head = head_chunk.ok_or_else(|| Error::Other("empty gRPC stream".into()))?;
+        if !head.has_message() {
+            return Err(box_err!("no raft message in the first chunk"));
+        }
+
+        let meta = head.take_message();
+        let snapshot = meta.get_message().get_snapshot();
+        let key = TabletSnapKey::from_region_snap(
+            meta.get_region_id(),
+            meta.get_to_peer().get_id(),
+            meta.get_message().get_snapshot(),
+        );
+
+        let data = snapshot.get_data();
+        let mut snapshot_data = RaftSnapshotData::default();
+        snapshot_data.merge_from_bytes(data)?;
+        let snapshot_meta = snapshot_data.take_meta();
+        let io_type = if snapshot_meta.get_for_balance() {
+            IoType::LoadBalance
+        } else {
+            IoType::Replication
+        };
+        let _with_io_type = WithIoType::new(io_type);
```

Comment:
This is a RAII style guard (i.e. the type is erased if `_with_io_type` goes out of scope). You need to save it in the context.

Reply:
fixed